cohttp-eio/src/client.ml

```diff
@@ -0,0 +1,115 @@
+  Buf_write.string writer (Http.Method.to_string meth);
+  Buf_write.char writer ' ';
+  Buf_write.string writer resource_path;
+  Buf_write.char writer ' ';
+  Buf_write.string writer (Http.Version.to_string version);
+  Buf_write.string writer "\r\n";
+  Rwer.write_headers writer headers;
+  Buf_write.string writer "\r\n";
+  Body.write_body writer body
+
+(* response parser *)
+
+let is_digit = function '0' .. '9' -> true | _ -> false
+
+let status_code =
+  let open Rwer in
+  let open Buf_read.Syntax in
+  let+ status = take_while1 is_digit in
+  Http.Status.of_int (int_of_string status)
+
+let reason_phrase =
+  Buf_read.take_while (function
+    | '\x21' .. '\x7E' | '\t' | ' ' -> true
+    | _ -> false)
+
+(* https://datatracker.ietf.org/doc/html/rfc7230#section-3.1.2 *)
+let response buf_read =
+  let open Buf_read.Syntax in
+  let version = Rwer.(version <* space) buf_read in
+  let status = Rwer.(status_code <* space) buf_read in
+  let () = Rwer.(reason_phrase *> crlf *> Buf_read.return ()) buf_read in
+  let headers = Rwer.http_headers buf_read in
+  Http.Response.make ~version ~status ~headers ()
+
+(* Generic HTTP call *)
+
+let call ?(meth = `GET) ?(version = `HTTP_1_1) ?(headers = Http.Header.init ())
+    ?(body = Body.Empty) ~conn host resource_path =
+  let host =
+    match host with
+    | host, Some port -> host ^ ":" ^ string_of_int port
+    | host, None -> host
+  in
+  Buf_write.with_flow ~initial_size:0x1000 conn (fun writer ->
+      let headers = Http.Header.add_unless_exists headers "Host" host in
+      write_request writer (meth, version, headers, resource_path, body);
+      let reader =
+        Eio.Buf_read.of_flow ~initial_size:0x1000 ~max_size:max_int conn
+      in
+      let response = response reader in
+      (response, reader))
+
+(*  HTTP Calls with Body Disallowed *)
+
+let get ?version ?headers ~conn host resource_path =
+  call ~meth:`GET ?version ?headers ~conn host resource_path
+
+let head ?version ?headers ~conn host resource_path =
+  call ~meth:`HEAD ?version ?headers ~conn host resource_path
+
+let delete ?version ?headers ~conn host resource_path =
+  call ~meth:`DELETE ?version ?headers ~conn host resource_path
+
+(*  HTTP Calls with Body Allowed *)
+
+let post ?version ?headers ?body ~conn host resource_path =
+  call ~meth:`POST ?version ?headers ?body ~conn host resource_path
+
+let put ?version ?headers ?body ~conn host resource_path =
+  call ~meth:`PUT ?version ?headers ?body ~conn host resource_path
+
+let patch ?version ?headers ?body ~conn host resource_path =
+  call ~meth:`PATCH ?version ?headers ?body ~conn host resource_path
+
+(* Response Body *)
+
+let read_fixed ((response, reader) : Http.Response.t * Buf_read.t) =
+  match
+    Http.Header.get response.headers "Content-Length"
+    |> Option.get
+    |> int_of_string
+  with
+  | content_length -> Buf_read.take content_length reader
+  | exception _ -> Buf_read.take_all reader
```

Comment:
Don't treat all errors (e.g. not-an-integer) as a missing content-length:
```suggestion
let read_fixed ((response, reader) : Http.Response.t * Buf_read.t) =
  match Http.Header.get response.headers "Content-Length" with
  | Some content_length -> Buf_read.take (int_of_string content_length) reader
  | None -> Buf_read.take_all reader
```

Reply:
Shouldn't an invalid `content-length` value(integer) be understood as an invalid `Content-Length` header and be treated as such? Specifically, 
>  5.  If a valid Content-Length header field is present without
       Transfer-Encoding, its decimal value defines the expected message
       body length in octets.  If the sender closes the connection or
       the recipient times out before the indicated number of octets are
       received, the recipient MUST consider the message to be
       incomplete and close the connection.

I assumed the "invalid" in a broader sense, not just missing 'Content-Length' header. So this would be invalid `Content-length: abc` since `abc` can't be a valid integer value in OCaml.

Reply:
That's an interesting point. https://www.rfc-editor.org/rfc/rfc7230#section-3.3.2 says:

> Any Content-Length field value greater than or equal to zero is valid.

I think it's assuming that the value will always be an integer, though it's not clear. It's also unclear why negative values should be ignored - possibly some systems use that to indicate streaming?

The older https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html says:

> When a Content-Length is given in a message where a message-body is allowed, its field value MUST exactly match the number of OCTETs in the message-body. HTTP/1.1 user agents MUST notify the user when an invalid length is received and detected.

Reply:
Thanks for the pointers. They have been very helpful. I have created issues based on our conversation so far.
https://github.com/mirage/ocaml-cohttp/issues/927
https://github.com/mirage/ocaml-cohttp/issues/926

I want to address them in PRs after this.